// devextreme/esm/__internal/core/license/types.js
/**
 * DevExtreme (esm/__internal/core/license/types.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
export var TokenKind;
! function(TokenKind) {
    TokenKind.corrupted = "corrupted";
    TokenKind.verified = "verified";
    TokenKind.internal = "internal"
}(TokenKind || (TokenKind = {}));
export const GENERAL_ERROR = {
    kind: TokenKind.corrupted,
    error: "general"
};
export const VERIFICATION_ERROR = {
    kind: TokenKind.corrupted,
    error: "verification"
};
export const DECODING_ERROR = {
    kind: TokenKind.corrupted,
    error: "decoding"
};
export const DESERIALIZATION_ERROR = {
    kind: TokenKind.corrupted,
    error: "deserialization"
};
export const PAYLOAD_ERROR = {
    kind: TokenKind.corrupted,
    error: "payload"
};
export const VERSION_ERROR = {
    kind: TokenKind.corrupted,
    error: "version"
};
export const PRODUCT_KIND_ERROR = {
    kind: TokenKind.corrupted,
    error: "product-kind"
};
