// devextreme/cjs/__internal/scheduler/shaders/current_time_shader_vertical.js
/**
 * DevExtreme (cjs/__internal/scheduler/shaders/current_time_shader_vertical.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _renderer = _interopRequireDefault(require("../../../core/renderer"));
var _size = require("../../../core/utils/size");
var _current_time_shader = _interopRequireDefault(require("./current_time_shader"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const DATE_TIME_SHADER_ALL_DAY_CLASS = "dx-scheduler-date-time-shader-all-day";
const DATE_TIME_SHADER_TOP_CLASS = "dx-scheduler-date-time-shader-top";
const DATE_TIME_SHADER_BOTTOM_CLASS = "dx-scheduler-date-time-shader-bottom";
class VerticalCurrentTimeShader extends _current_time_shader.default {
    renderShader() {
        let shaderHeight = this.getShaderHeight();
        const maxHeight = this.getShaderMaxHeight();
        const isSolidShader = shaderHeight > maxHeight;
        if (shaderHeight > maxHeight) {
            shaderHeight = maxHeight
        }(0, _size.setHeight)(this._$shader, shaderHeight);
        const groupCount = this._workSpace._getGroupCount() || 1;
        if (this._workSpace.isGroupedByDate()) {
            this.renderGroupedByDateShaderParts(groupCount, shaderHeight, maxHeight, isSolidShader)
        } else {
            this.renderShaderParts(groupCount, shaderHeight, maxHeight, isSolidShader)
        }
    }
    renderShaderParts(groupCount, shaderHeight, maxHeight, isSolidShader) {
        for (let i = 0; i < groupCount; i += 1) {
            const shaderWidth = this.getShaderWidth();
            this.renderTopShader(this._$shader, shaderHeight, shaderWidth, i);
            if (!isSolidShader) {
                this.renderBottomShader(this._$shader, maxHeight, shaderHeight, shaderWidth, i)
            }
            this.renderAllDayShader(shaderWidth, i)
        }
    }
    renderGroupedByDateShaderParts(groupCount, shaderHeight, maxHeight, isSolidShader) {
        const shaderWidth = this.getShaderWidth();
        const bottomShaderWidth = shaderHeight < 0 ? shaderWidth : shaderWidth - this._workSpace.getCellWidth();
        const normalizedShaderHeight = Math.max(shaderHeight, 0);
        this.renderTopShader(this._$shader, normalizedShaderHeight, shaderWidth * groupCount, 0);
        if (!isSolidShader) {
            this.renderBottomShader(this._$shader, maxHeight, normalizedShaderHeight, bottomShaderWidth * groupCount + this._workSpace.getCellWidth(), 0)
        }
        this.renderAllDayShader(shaderWidth * groupCount, 0)
    }
    renderTopShader($shader, height, width, i) {
        this._$topShader = (0, _renderer.default)("<div>").addClass(DATE_TIME_SHADER_TOP_CLASS);
        if (width) {
            (0, _size.setWidth)(this._$topShader, width)
        }
        if (height) {
            (0, _size.setHeight)(this._$topShader, height)
        }
        this._$topShader.css("marginTop", this.getShaderTopOffset(i));
        this._$topShader.css("left", this.getShaderOffset(i, width));
        $shader.append(this._$topShader)
    }
    renderBottomShader($shader, maxHeight, height, width, i) {
        this._$bottomShader = (0, _renderer.default)("<div>").addClass(DATE_TIME_SHADER_BOTTOM_CLASS);
        const shaderWidth = height < 0 ? width : width - this._workSpace.getCellWidth();
        const shaderHeight = height < 0 ? maxHeight : maxHeight - height;
        (0, _size.setWidth)(this._$bottomShader, shaderWidth);
        (0, _size.setHeight)(this._$bottomShader, shaderHeight);
        this._$bottomShader.css("left", this.getShaderOffset(i, width - this._workSpace.getCellWidth()));
        $shader.append(this._$bottomShader)
    }
    renderAllDayShader(shaderWidth, i) {
        if (this._workSpace.option("showAllDayPanel")) {
            this._$allDayIndicator = (0, _renderer.default)("<div>").addClass(DATE_TIME_SHADER_ALL_DAY_CLASS);
            (0, _size.setHeight)(this._$allDayIndicator, this._workSpace.getAllDayHeight());
            (0, _size.setWidth)(this._$allDayIndicator, shaderWidth);
            this._$allDayIndicator.css("left", this.getShaderOffset(i, shaderWidth));
            this._workSpace._$allDayPanel.prepend(this._$allDayIndicator)
        }
    }
    getShaderOffset(i, width) {
        return this._workSpace.getGroupedStrategy().getShaderOffset(i, width)
    }
    getShaderTopOffset(i) {
        return this._workSpace.getGroupedStrategy().getShaderTopOffset(i)
    }
    getShaderHeight() {
        return this._workSpace.getGroupedStrategy().getShaderHeight()
    }
    getShaderMaxHeight() {
        return this._workSpace.getGroupedStrategy().getShaderMaxHeight()
    }
    getShaderWidth() {
        return this._workSpace.getGroupedStrategy().getShaderWidth()
    }
    clean() {
        var _this$_workSpace;
        super.clean();
        if (null !== (_this$_workSpace = this._workSpace) && void 0 !== _this$_workSpace && _this$_workSpace._$allDayPanel) {
            this._workSpace._$allDayPanel.find(`.${DATE_TIME_SHADER_ALL_DAY_CLASS}`).remove()
        }
    }
}
var _default = exports.default = VerticalCurrentTimeShader;
