// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/model/cell/header_cell.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/model/cell/header_cell.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
const SELECTORS = {
    headerContent: "text-content",
    alignmentRight: "dx-text-content-alignment-right",
    alignmentLeft: "dx-text-content-alignment-left"
};
export class HeaderCellModel {
    constructor(root, addWidgetPrefix) {
        this.root = root;
        this.addWidgetPrefix = addWidgetPrefix
    }
    getElement() {
        return this.root
    }
    getText() {
        var _this$getHeaderConten;
        return (null === (_this$getHeaderConten = this.getHeaderContent()) || void 0 === _this$getHeaderConten ? void 0 : _this$getHeaderConten.textContent) ?? ""
    }
    getHeaderContent() {
        var _this$root;
        return (null === (_this$root = this.root) || void 0 === _this$root ? void 0 : _this$root.querySelector(`.${this.addWidgetPrefix(SELECTORS.headerContent)}`)) ?? null
    }
    getAlignment() {
        const content = this.getHeaderContent();
        const alignedRight = null === content || void 0 === content ? void 0 : content.classList.contains(SELECTORS.alignmentRight);
        const alignedLeft = null === content || void 0 === content ? void 0 : content.classList.contains(SELECTORS.alignmentLeft);
        const alignedCenter = alignedLeft && alignedRight;
        switch (true) {
            case alignedCenter:
                return "center";
            case alignedRight:
                return "right";
            default:
                return "left"
        }
    }
}
