// devextreme/cjs/__internal/scheduler/view_model/filtration/utils/add_all_day_panel_occupation.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/filtration/utils/add_all_day_panel_occupation.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addAllDayPanelOccupation = void 0;
var _base = require("../../../r1/utils/base");
const addAllDayPanelOccupation = (appointments, _ref) => {
    let {
        supportAllDayPanel: supportAllDayPanel,
        allDayPanelMode: allDayPanelMode
    } = _ref;
    return appointments.map((appointment => {
        const isAllDayPanelOccupied = supportAllDayPanel && (0, _base.isAppointmentTakesAllDay)({
            allDay: appointment.allDay,
            startDate: new Date(appointment.source.startDate),
            endDate: new Date(appointment.source.endDate)
        }, allDayPanelMode);
        return Object.assign({}, appointment, {
            isAllDayPanelOccupied: isAllDayPanelOccupied
        })
    }))
};
exports.addAllDayPanelOccupation = addAllDayPanelOccupation;
