// devextreme/esm/__internal/grids/new/grid_core/filtering/filter_panel/options.js
/**
 * DevExtreme (esm/__internal/grids/new/grid_core/filtering/filter_panel/options.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
export const defaultOptions = {
    filterBuilder: {
        groupOperationDescriptions: {
            and: void 0,
            or: void 0,
            notAnd: void 0,
            notOr: void 0
        },
        filterOperationDescriptions: {
            between: void 0,
            equal: void 0,
            notEqual: void 0,
            lessThan: void 0,
            lessThanOrEqual: void 0,
            greaterThan: void 0,
            greaterThanOrEqual: void 0,
            startsWith: void 0,
            contains: void 0,
            notContains: void 0,
            endsWith: void 0,
            isBlank: void 0,
            isNotBlank: void 0
        }
    },
    filterPanel: {
        visible: false,
        filterEnabled: true,
        texts: {
            createFilter: void 0,
            clearFilter: void 0,
            filterEnabledHint: void 0
        }
    },
    filterBuilderPopup: {}
};
