// devextreme/esm/__internal/scheduler/__tests__/__mock__/create_appointment_popup.js
/**
 * DevExtreme (esm/__internal/scheduler/__tests__/__mock__/create_appointment_popup.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    jest
} from "@jest/globals";
import $ from "../../../../core/renderer";
import {
    Deferred
} from "../../../../core/utils/deferred";
import {
    mockTimeZoneCalculator
} from "../../__mock__/timezone_calculator.mock";
import {
    AppointmentForm
} from "../../appointment_popup/m_form";
import {
    ACTION_TO_APPOINTMENT,
    APPOINTMENT_POPUP_CLASS,
    AppointmentPopup
} from "../../appointment_popup/m_popup";
import {
    AppointmentDataAccessor
} from "../../utils/data_accessor/appointment_data_accessor";
import {
    ResourceManager
} from "../../utils/resource_manager/resource_manager";
import {
    PopupModel
} from "./model/popup";
const DEFAULT_FIELDS = {
    startDateExpr: "startDate",
    endDateExpr: "endDate",
    startDateTimeZoneExpr: "startDateTimeZone",
    endDateTimeZoneExpr: "endDateTimeZone",
    allDayExpr: "allDay",
    textExpr: "text",
    descriptionExpr: "description",
    recurrenceRuleExpr: "recurrenceRule",
    recurrenceExceptionExpr: "recurrenceException",
    disabledExpr: "disabled",
    visibleExpr: "visible"
};
const DEFAULT_EDITING = {
    allowAdding: true,
    allowUpdating: true,
    allowDeleting: true,
    allowResizing: true,
    allowDragging: true,
    legacyForm: false
};
const DEFAULT_APPOINTMENT = {
    text: "Test Appointment",
    startDate: new Date(2021, 3, 26, 9, 30),
    endDate: new Date(2021, 3, 26, 11, 0)
};
const resolvedDeferred = () => {
    const d = new Deferred;
    d.resolve();
    return d.promise()
};

function createComponent(element, Component, opts) {
    return new Component($(element), opts)
}
const disposables = [];
export const disposeAppointmentPopups = () => {
    disposables.forEach((fn => fn()));
    disposables.length = 0;
    document.body.innerHTML = ""
};
export const createAppointmentPopup = async function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    const container = document.createElement("div");
    document.body.appendChild(container);
    const dataAccessors = new AppointmentDataAccessor(DEFAULT_FIELDS, false);
    const resourceManager = new ResourceManager([]);
    const timeZoneCalculator = mockTimeZoneCalculator;
    const editing = Object.assign({}, DEFAULT_EDITING, options.editing);
    const addAppointment = options.addAppointment ?? jest.fn(resolvedDeferred);
    const updateAppointment = options.updateAppointment ?? jest.fn(resolvedDeferred);
    const focus = jest.fn();
    const updateScrollPosition = jest.fn();
    const formSchedulerProxy = {
        getResourceById: () => resourceManager.resourceById,
        getDataAccessors: () => dataAccessors,
        createComponent: createComponent,
        getEditingConfig: () => editing,
        getResourceManager: () => resourceManager,
        getFirstDayOfWeek: () => options.firstDayOfWeek ?? 0,
        getStartDayHour: () => options.startDayHour ?? 0,
        getCalculatedEndDate: startDate => {
            const endDate = new Date(startDate);
            endDate.setHours(endDate.getHours() + 1);
            return endDate
        },
        getTimeZoneCalculator: () => timeZoneCalculator
    };
    const form = new AppointmentForm(formSchedulerProxy);
    const noop = () => {};
    const popupSchedulerProxy = {
        getElement: () => $(container),
        createComponent: createComponent,
        focus: focus,
        getResourceManager: () => resourceManager,
        getEditingConfig: () => editing,
        getTimeZoneCalculator: () => timeZoneCalculator,
        getDataAccessors: () => dataAccessors,
        getAppointmentFormOpening: () => options.onAppointmentFormOpening ?? noop,
        processActionResult: (arg, callback) => {
            callback(arg.cancel)
        },
        addAppointment: addAppointment,
        updateAppointment: updateAppointment,
        updateScrollPosition: updateScrollPosition
    };
    const popup = new AppointmentPopup(popupSchedulerProxy, form);
    const appointmentData = options.appointmentData ?? Object.assign({}, DEFAULT_APPOINTMENT);
    const action = options.action ?? ACTION_TO_APPOINTMENT.CREATE;
    popup.show(appointmentData, {
        action: action,
        allowSaving: true
    });
    await new Promise(process.nextTick);
    const selector = `.dx-overlay-wrapper.${APPOINTMENT_POPUP_CLASS}`;
    const overlayWrapper = document.querySelector(selector);
    if (!overlayWrapper) {
        throw new Error("AppointmentPopup overlay wrapper not found in DOM")
    }
    const POM = new PopupModel(overlayWrapper);
    const dispose = () => {
        popup.dispose();
        container.remove()
    };
    disposables.push(dispose);
    return {
        container: container,
        popup: popup,
        form: form,
        POM: POM,
        callbacks: {
            addAppointment: addAppointment,
            updateAppointment: updateAppointment,
            focus: focus,
            updateScrollPosition: updateScrollPosition
        },
        dispose: dispose
    }
};
