// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/model/filter_panel.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/model/filter_panel.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.FilterPanelModel = void 0;
var _checkbox = require("../../../../../ui/__tests__/__mock__/model/checkbox");
class FilterPanelModel {
    constructor(root, prefix) {
        this.filterPanelClass = `dx-${prefix}-filter-panel`;
        this.filterPanelTextClass = `dx-${prefix}-filter-panel-text`;
        this.filterPanelClearFilterClass = `dx-${prefix}-filter-panel-clear-filter`;
        this.filterPanelCheckboxClass = `dx-${prefix}-filter-panel-checkbox`;
        this.root = root
    }
    getElement() {
        return this.root.querySelector(`.${this.filterPanelClass}`)
    }
    getCreateFilterButton() {
        return this.getElement().querySelector(`.${this.filterPanelTextClass}`)
    }
    getClearFilterButton() {
        return this.getElement().querySelector(`.${this.filterPanelClearFilterClass}`)
    }
    getEnableFilterCheckbox() {
        const checkboxElement = this.getElement().querySelector(`.${this.filterPanelCheckboxClass}`);
        return new _checkbox.CheckBoxModel(checkboxElement)
    }
    isVisible() {
        const element = this.getElement();
        return null !== element && "none" !== element.style.display
    }
}
exports.FilterPanelModel = FilterPanelModel;
