// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GridCoreModel = void 0;
var _renderer = _interopRequireDefault(require("../../../../../../core/renderer"));
var _filter_builder = require("../../../../../filter_builder/__tests__/__mock__/model/filter_builder");
var _pager = require("../../../../../pagination/__tests__/__mock__/model/pager");
var _load_panel = require("../../../../../ui/__tests__/__mock__/model/load_panel");
var _toast = require("../../../../../ui/__tests__/__mock__/model/toast");
var _ai_prompt_editor = require("./ai_prompt_editor");
var _ai_header_cell = require("./cell/ai_header_cell");
var _data_cell = require("./cell/data_cell");
var _header_cell = require("./cell/header_cell");
var _column_chooser = require("./column_chooser");
var _confirmation_dialog = require("./confirmation_dialog");
var _edit_form = require("./edit_form");
var _filter_panel = require("./filter_panel");
var _data_row = require("./row/data_row");
var _group_row = require("./row/group_row");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const SELECTORS = {
    headerRowClass: "dx-header-row",
    dataRowClass: "dx-data-row",
    groupRowClass: "dx-group-row",
    aiDialog: "dx-aidialog",
    aiPromptEditor: "dx-ai-prompt-editor",
    toast: "dx-toast",
    loadPanel: "dx-loadpanel",
    editForm: "edit-form",
    headerCellIndicators: "dx-column-indicators",
    headerCellFilter: "dx-header-filter",
    revertButton: "dx-revert-button"
};
class GridCoreModel {
    constructor(root) {
        this.root = root
    }
    getWidgetName() {
        return this.NAME.slice(2).toLowerCase()
    }
    getFilterPanelPrefix() {
        return this.getWidgetName()
    }
    getPromptEditorContainer() {
        return this.root.querySelector(`.${SELECTORS.aiPromptEditor}`)
    }
    getHeaderCells() {
        return this.root.querySelectorAll(`.${SELECTORS.headerRowClass} > td`)
    }
    getHeaderCell(columnIndex) {
        return new _header_cell.HeaderCellModel(this.getHeaderCells()[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getAIHeaderCell(columnIndex) {
        return new _ai_header_cell.AIHeaderCellModel(this.getHeaderCells()[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getHeaderCellFilter(columnIndex) {
        const $headerCell = (0, _renderer.default)(this.getHeaderCells()[columnIndex]);
        const headerFilterSelector = `.${SELECTORS.headerCellIndicators} > .${SELECTORS.headerCellFilter}`;
        return $headerCell.find(headerFilterSelector)
    }
    getDataRows() {
        return this.root.querySelectorAll(`.${SELECTORS.dataRowClass}`)
    }
    getDataRow(rowIndex) {
        return new _data_row.DataRowModel(this.getDataRows()[rowIndex])
    }
    getDataCells(rowIndex) {
        return this.root.querySelectorAll(`.${SELECTORS.dataRowClass}:nth-child(${rowIndex+1}) > td`)
    }
    getDataCell(rowIndex, columnIndex) {
        return new _data_cell.DataCellModel(this.getDataCells(rowIndex)[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getGroupRows() {
        return this.root.querySelectorAll(`.${SELECTORS.groupRowClass}`)
    }
    getGroupRow(rowIndex) {
        return new _group_row.GroupRowModel(this.getGroupRows()[rowIndex])
    }
    getHeaderByText(text) {
        return (0, _renderer.default)(Array.from(this.getHeaderCells()).find((el => (0, _renderer.default)(el).text().includes(text))))
    }
    getAIDialog() {
        return document.body.querySelector(`.${SELECTORS.aiDialog}`)
    }
    getAIPromptEditor() {
        return new _ai_prompt_editor.AIPromptEditorModel(this.getPromptEditorContainer())
    }
    getToastContainer() {
        return document.body.querySelector(`.${SELECTORS.toast}`)
    }
    getToast() {
        return new _toast.ToastModel(this.getToastContainer())
    }
    getRevertButton() {
        return document.body.querySelector(`.${SELECTORS.revertButton}`)
    }
    addWidgetPrefix(classNames) {
        const widgetName = this.getWidgetName();
        return `dx-${widgetName}${classNames?`-${classNames}`:""}`
    }
    getLoadPanel() {
        return new _load_panel.LoadPanelModel(document.body.querySelector(`.${SELECTORS.loadPanel}`))
    }
    getEditForm() {
        return new _edit_form.EditFormModel(this.root.querySelector(`.${this.addWidgetPrefix(SELECTORS.editForm)}`))
    }
    getColumnChooser() {
        return new _column_chooser.ColumnChooserModel(this.getWidgetName())
    }
    getFilterPanel() {
        return new _filter_panel.FilterPanelModel(this.root, this.getFilterPanelPrefix())
    }
    getFilterBuilder() {
        return new _filter_builder.FilterBuilderModel
    }
    getPager() {
        return new _pager.PagerModel(this.root)
    }
    getConfirmationDialog() {
        return new _confirmation_dialog.ConfirmationDialogModel
    }
    apiOption(name, value) {
        const instance = this.getInstance();
        if (1 === arguments.length) {
            return instance.option(name)
        }
        instance.option(name, value);
        return
    }
}
exports.GridCoreModel = GridCoreModel;
