// devextreme/cjs/__internal/scheduler/__tests__/__mock__/create_appointment_popup.js
/**
 * DevExtreme (cjs/__internal/scheduler/__tests__/__mock__/create_appointment_popup.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.disposeAppointmentPopups = exports.createAppointmentPopup = void 0;
var _globals = require("@jest/globals");
var _renderer = _interopRequireDefault(require("../../../../core/renderer"));
var _deferred = require("../../../../core/utils/deferred");
var _timezone_calculator = require("../../__mock__/timezone_calculator.mock");
var _m_form = require("../../appointment_popup/m_form");
var _m_popup = require("../../appointment_popup/m_popup");
var _appointment_data_accessor = require("../../utils/data_accessor/appointment_data_accessor");
var _resource_manager = require("../../utils/resource_manager/resource_manager");
var _popup = require("./model/popup");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const DEFAULT_FIELDS = {
    startDateExpr: "startDate",
    endDateExpr: "endDate",
    startDateTimeZoneExpr: "startDateTimeZone",
    endDateTimeZoneExpr: "endDateTimeZone",
    allDayExpr: "allDay",
    textExpr: "text",
    descriptionExpr: "description",
    recurrenceRuleExpr: "recurrenceRule",
    recurrenceExceptionExpr: "recurrenceException",
    disabledExpr: "disabled",
    visibleExpr: "visible"
};
const DEFAULT_EDITING = {
    allowAdding: true,
    allowUpdating: true,
    allowDeleting: true,
    allowResizing: true,
    allowDragging: true,
    legacyForm: false
};
const DEFAULT_APPOINTMENT = {
    text: "Test Appointment",
    startDate: new Date(2021, 3, 26, 9, 30),
    endDate: new Date(2021, 3, 26, 11, 0)
};
const resolvedDeferred = () => {
    const d = new _deferred.Deferred;
    d.resolve();
    return d.promise()
};

function createComponent(element, Component, opts) {
    return new Component((0, _renderer.default)(element), opts)
}
const disposables = [];
const disposeAppointmentPopups = () => {
    disposables.forEach((fn => fn()));
    disposables.length = 0;
    document.body.innerHTML = ""
};
exports.disposeAppointmentPopups = disposeAppointmentPopups;
const createAppointmentPopup = async function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    const container = document.createElement("div");
    document.body.appendChild(container);
    const dataAccessors = new _appointment_data_accessor.AppointmentDataAccessor(DEFAULT_FIELDS, false);
    const resourceManager = new _resource_manager.ResourceManager([]);
    const timeZoneCalculator = _timezone_calculator.mockTimeZoneCalculator;
    const editing = Object.assign({}, DEFAULT_EDITING, options.editing);
    const addAppointment = options.addAppointment ?? _globals.jest.fn(resolvedDeferred);
    const updateAppointment = options.updateAppointment ?? _globals.jest.fn(resolvedDeferred);
    const focus = _globals.jest.fn();
    const updateScrollPosition = _globals.jest.fn();
    const formSchedulerProxy = {
        getResourceById: () => resourceManager.resourceById,
        getDataAccessors: () => dataAccessors,
        createComponent: createComponent,
        getEditingConfig: () => editing,
        getResourceManager: () => resourceManager,
        getFirstDayOfWeek: () => options.firstDayOfWeek ?? 0,
        getStartDayHour: () => options.startDayHour ?? 0,
        getCalculatedEndDate: startDate => {
            const endDate = new Date(startDate);
            endDate.setHours(endDate.getHours() + 1);
            return endDate
        },
        getTimeZoneCalculator: () => timeZoneCalculator
    };
    const form = new _m_form.AppointmentForm(formSchedulerProxy);
    const noop = () => {};
    const popupSchedulerProxy = {
        getElement: () => (0, _renderer.default)(container),
        createComponent: createComponent,
        focus: focus,
        getResourceManager: () => resourceManager,
        getEditingConfig: () => editing,
        getTimeZoneCalculator: () => timeZoneCalculator,
        getDataAccessors: () => dataAccessors,
        getAppointmentFormOpening: () => options.onAppointmentFormOpening ?? noop,
        processActionResult: (arg, callback) => {
            callback(arg.cancel)
        },
        addAppointment: addAppointment,
        updateAppointment: updateAppointment,
        updateScrollPosition: updateScrollPosition
    };
    const popup = new _m_popup.AppointmentPopup(popupSchedulerProxy, form);
    const appointmentData = options.appointmentData ?? Object.assign({}, DEFAULT_APPOINTMENT);
    const action = options.action ?? _m_popup.ACTION_TO_APPOINTMENT.CREATE;
    popup.show(appointmentData, {
        action: action,
        allowSaving: true
    });
    await new Promise(process.nextTick);
    const selector = `.dx-overlay-wrapper.${_m_popup.APPOINTMENT_POPUP_CLASS}`;
    const overlayWrapper = document.querySelector(selector);
    if (!overlayWrapper) {
        throw new Error("AppointmentPopup overlay wrapper not found in DOM")
    }
    const POM = new _popup.PopupModel(overlayWrapper);
    const dispose = () => {
        popup.dispose();
        container.remove()
    };
    disposables.push(dispose);
    return {
        container: container,
        popup: popup,
        form: form,
        POM: POM,
        callbacks: {
            addAppointment: addAppointment,
            updateAppointment: updateAppointment,
            focus: focus,
            updateScrollPosition: updateScrollPosition
        },
        dispose: dispose
    }
};
exports.createAppointmentPopup = createAppointmentPopup;
