// devextreme/cjs/__internal/ui/__tests__/__mock__/model/text_area.js
/**
 * DevExtreme (cjs/__internal/ui/__tests__/__mock__/model/text_area.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TextAreaModel = void 0;
var _m_text_area = _interopRequireDefault(require("../../../../ui/m_text_area"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const CLASSES = {
    textArea: "dx-texteditor-input"
};
class TextAreaModel {
    constructor(root) {
        this.root = root
    }
    getElement() {
        return this.root
    }
    getInputElement() {
        return this.root.querySelector(`.${CLASSES.textArea}`)
    }
    setValue(value) {
        var _this$getInstance;
        null === (_this$getInstance = this.getInstance()) || void 0 === _this$getInstance || _this$getInstance.option("value", value)
    }
    getInstance() {
        return _m_text_area.default.getInstance(this.root)
    }
}
exports.TextAreaModel = TextAreaModel;
