// devextreme/cjs/__internal/core/license/const.js
/**
 * DevExtreme (cjs/__internal/core/license/const.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.SUBSCRIPTION_NAMES = exports.RTM_MIN_PATCH_VERSION = exports.NBSP = exports.LICENSING_DOC_LINK = exports.KEY_SPLITTER = exports.FORMAT = exports.BUY_NOW_LINK = void 0;
const FORMAT = exports.FORMAT = 1;
const RTM_MIN_PATCH_VERSION = exports.RTM_MIN_PATCH_VERSION = 3;
const KEY_SPLITTER = exports.KEY_SPLITTER = ".";
const BUY_NOW_LINK = exports.BUY_NOW_LINK = "https://go.devexpress.com/Licensing_Installer_Watermark_DevExtremeJQuery.aspx";
const LICENSING_DOC_LINK = exports.LICENSING_DOC_LINK = "https://go.devexpress.com/Licensing_Documentation_DevExtremeJQuery.aspx";
const NBSP = exports.NBSP = "\xa0";
const SUBSCRIPTION_NAMES = exports.SUBSCRIPTION_NAMES = `Universal, DXperience, ASP.NET${NBSP}and${NBSP}Blazor, DevExtreme${NBSP}Complete`;
