// devextreme/cjs/__internal/core/license/types.js
/**
 * DevExtreme (cjs/__internal/core/license/types.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.VERSION_ERROR = exports.VERIFICATION_ERROR = exports.TokenKind = exports.PRODUCT_KIND_ERROR = exports.PAYLOAD_ERROR = exports.GENERAL_ERROR = exports.DESERIALIZATION_ERROR = exports.DECODING_ERROR = void 0;
var TokenKind;
! function(TokenKind) {
    TokenKind.corrupted = "corrupted";
    TokenKind.verified = "verified";
    TokenKind.internal = "internal"
}(TokenKind || (exports.TokenKind = TokenKind = {}));
const GENERAL_ERROR = exports.GENERAL_ERROR = {
    kind: TokenKind.corrupted,
    error: "general"
};
const VERIFICATION_ERROR = exports.VERIFICATION_ERROR = {
    kind: TokenKind.corrupted,
    error: "verification"
};
const DECODING_ERROR = exports.DECODING_ERROR = {
    kind: TokenKind.corrupted,
    error: "decoding"
};
const DESERIALIZATION_ERROR = exports.DESERIALIZATION_ERROR = {
    kind: TokenKind.corrupted,
    error: "deserialization"
};
const PAYLOAD_ERROR = exports.PAYLOAD_ERROR = {
    kind: TokenKind.corrupted,
    error: "payload"
};
const VERSION_ERROR = exports.VERSION_ERROR = {
    kind: TokenKind.corrupted,
    error: "version"
};
const PRODUCT_KIND_ERROR = exports.PRODUCT_KIND_ERROR = {
    kind: TokenKind.corrupted,
    error: "product-kind"
};
