// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/model/edit_form.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/model/edit_form.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    PopupModel
} from "../../../../../ui/__tests__/__mock__/model/popup";
const SELECTORS = {
    textEditorInput: "dx-texteditor-input",
    item: "dx-item",
    toolbar: "dx-toolbar",
    saveButton: '[aria-label*="Save"]'
};
export class EditFormModel extends PopupModel {
    constructor(root) {
        super();
        this.root = root
    }
    getSaveButton() {
        const overlay = this.getOverlayContent();
        const toolbar = null === overlay || void 0 === overlay ? void 0 : overlay.querySelector(`.${SELECTORS.toolbar}`);
        return null === toolbar || void 0 === toolbar ? void 0 : toolbar.querySelector(SELECTORS.saveButton)
    }
    getItem(id) {
        var _this$root;
        return (null === (_this$root = this.root) || void 0 === _this$root ? void 0 : _this$root.querySelector(`.${SELECTORS.textEditorInput}[id*=_${id}]`)) ?? null
    }
    getItems() {
        var _this$root2;
        return (null === (_this$root2 = this.root) || void 0 === _this$root2 ? void 0 : _this$root2.querySelectorAll(`.${SELECTORS.item}`)) ?? null
    }
}
