// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import $ from "../../../../../../core/renderer";
import {
    FilterBuilderModel
} from "../../../../../filter_builder/__tests__/__mock__/model/filter_builder";
import {
    PagerModel
} from "../../../../../pagination/__tests__/__mock__/model/pager";
import {
    LoadPanelModel
} from "../../../../../ui/__tests__/__mock__/model/load_panel";
import {
    ToastModel
} from "../../../../../ui/__tests__/__mock__/model/toast";
import {
    AIPromptEditorModel
} from "./ai_prompt_editor";
import {
    AIHeaderCellModel
} from "./cell/ai_header_cell";
import {
    DataCellModel
} from "./cell/data_cell";
import {
    HeaderCellModel
} from "./cell/header_cell";
import {
    ColumnChooserModel
} from "./column_chooser";
import {
    ConfirmationDialogModel
} from "./confirmation_dialog";
import {
    EditFormModel
} from "./edit_form";
import {
    FilterPanelModel
} from "./filter_panel";
import {
    DataRowModel
} from "./row/data_row";
import {
    GroupRowModel
} from "./row/group_row";
const SELECTORS = {
    headerRowClass: "dx-header-row",
    dataRowClass: "dx-data-row",
    groupRowClass: "dx-group-row",
    aiDialog: "dx-aidialog",
    aiPromptEditor: "dx-ai-prompt-editor",
    toast: "dx-toast",
    loadPanel: "dx-loadpanel",
    editForm: "edit-form",
    headerCellIndicators: "dx-column-indicators",
    headerCellFilter: "dx-header-filter",
    revertButton: "dx-revert-button"
};
export class GridCoreModel {
    constructor(root) {
        this.root = root
    }
    getWidgetName() {
        return this.NAME.slice(2).toLowerCase()
    }
    getFilterPanelPrefix() {
        return this.getWidgetName()
    }
    getPromptEditorContainer() {
        return this.root.querySelector(`.${SELECTORS.aiPromptEditor}`)
    }
    getHeaderCells() {
        return this.root.querySelectorAll(`.${SELECTORS.headerRowClass} > td`)
    }
    getHeaderCell(columnIndex) {
        return new HeaderCellModel(this.getHeaderCells()[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getAIHeaderCell(columnIndex) {
        return new AIHeaderCellModel(this.getHeaderCells()[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getHeaderCellFilter(columnIndex) {
        const $headerCell = $(this.getHeaderCells()[columnIndex]);
        const headerFilterSelector = `.${SELECTORS.headerCellIndicators} > .${SELECTORS.headerCellFilter}`;
        return $headerCell.find(headerFilterSelector)
    }
    getDataRows() {
        return this.root.querySelectorAll(`.${SELECTORS.dataRowClass}`)
    }
    getDataRow(rowIndex) {
        return new DataRowModel(this.getDataRows()[rowIndex])
    }
    getDataCells(rowIndex) {
        return this.root.querySelectorAll(`.${SELECTORS.dataRowClass}:nth-child(${rowIndex+1}) > td`)
    }
    getDataCell(rowIndex, columnIndex) {
        return new DataCellModel(this.getDataCells(rowIndex)[columnIndex], this.addWidgetPrefix.bind(this))
    }
    getGroupRows() {
        return this.root.querySelectorAll(`.${SELECTORS.groupRowClass}`)
    }
    getGroupRow(rowIndex) {
        return new GroupRowModel(this.getGroupRows()[rowIndex])
    }
    getHeaderByText(text) {
        return $(Array.from(this.getHeaderCells()).find((el => $(el).text().includes(text))))
    }
    getAIDialog() {
        return document.body.querySelector(`.${SELECTORS.aiDialog}`)
    }
    getAIPromptEditor() {
        return new AIPromptEditorModel(this.getPromptEditorContainer())
    }
    getToastContainer() {
        return document.body.querySelector(`.${SELECTORS.toast}`)
    }
    getToast() {
        return new ToastModel(this.getToastContainer())
    }
    getRevertButton() {
        return document.body.querySelector(`.${SELECTORS.revertButton}`)
    }
    addWidgetPrefix(classNames) {
        const widgetName = this.getWidgetName();
        return `dx-${widgetName}${classNames?`-${classNames}`:""}`
    }
    getLoadPanel() {
        return new LoadPanelModel(document.body.querySelector(`.${SELECTORS.loadPanel}`))
    }
    getEditForm() {
        return new EditFormModel(this.root.querySelector(`.${this.addWidgetPrefix(SELECTORS.editForm)}`))
    }
    getColumnChooser() {
        return new ColumnChooserModel(this.getWidgetName())
    }
    getFilterPanel() {
        return new FilterPanelModel(this.root, this.getFilterPanelPrefix())
    }
    getFilterBuilder() {
        return new FilterBuilderModel
    }
    getPager() {
        return new PagerModel(this.root)
    }
    getConfirmationDialog() {
        return new ConfirmationDialogModel
    }
    apiOption(name, value) {
        const instance = this.getInstance();
        if (1 === arguments.length) {
            return instance.option(name)
        }
        instance.option(name, value);
        return
    }
}
