// devextreme/esm/__internal/scheduler/view_model/filtration/utils/get_filter_options/get_filter_options.test.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/filtration/utils/get_filter_options/get_filter_options.test.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    describe,
    expect,
    it
} from "@jest/globals";
import {
    getCompareOptions
} from "../../../../../scheduler/view_model/common/get_compare_options";
import {
    mockAppointmentDataAccessor
} from "../../../../__mock__/appointment_data_accessor.mock";
import {
    ResourceManager
} from "../../../../utils/resource_manager/resource_manager";
import {
    getFilterOptions
} from "./get_filter_options";
export const getSchedulerMock = _ref => {
    let {
        type: type,
        startDayHour: startDayHour,
        endDayHour: endDayHour,
        offsetMinutes: offsetMinutes,
        resourceManager: resourceManager,
        dateRange: dateRange
    } = _ref;
    return {
        currentView: {
            type: type,
            skippedDays: []
        },
        getWorkSpace: () => ({
            getDateRange: () => dateRange ?? [new Date(2e3, 0, 10, startDayHour), new Date(2e3, 0, 11, endDayHour)]
        }),
        getTimeZone: () => "Etc/UTC",
        getViewOption: name => ({
            startDayHour: startDayHour,
            endDayHour: endDayHour,
            allDayPanelMode: "allDay"
        } [name]),
        option: name => ({
            firstDayOfWeek: 0,
            showAllDayPanel: true
        } [name]),
        getViewOffsetMs: () => 6e4 * offsetMinutes,
        resourceManager: resourceManager ?? new ResourceManager([]),
        _dataAccessors: mockAppointmentDataAccessor
    }
};
describe("getFilterOptions", (() => {
    ["agenda", "month"].forEach((type => {
        it(`should return correct filter options for ${type} view`, (() => {
            const schedulerStore = getSchedulerMock({
                type: type,
                startDayHour: 0,
                endDayHour: 24,
                offsetMinutes: 30
            });
            const compareOptions = getCompareOptions(schedulerStore);
            const filterOptions = getFilterOptions(schedulerStore, compareOptions);
            expect(filterOptions).toEqual({
                allDayPanelMode: "allDay",
                showAllDayPanel: true,
                supportAllDayPanel: false,
                isDateTimeView: false,
                resourceManager: schedulerStore.resourceManager,
                dataAccessor: schedulerStore._dataAccessors,
                timeZone: "Etc/UTC",
                firstDayOfWeek: 0,
                allDayIntervals: [{
                    min: Date.UTC(2e3, 0, 10, 0, 30),
                    max: Date.UTC(2e3, 0, 12, 0, 30)
                }],
                regularIntervals: [{
                    min: Date.UTC(2e3, 0, 10, 0, 30),
                    max: Date.UTC(2e3, 0, 12, 0, 30)
                }]
            })
        }))
    }))
}));
