// devextreme/cjs/__internal/filter_builder/__tests__/__mock__/model/filter_builder.js
/**
 * DevExtreme (cjs/__internal/filter_builder/__tests__/__mock__/model/filter_builder.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.FilterBuilderModel = void 0;
var _popup = require("../../../../ui/__tests__/__mock__/model/popup");
var _tree_view = require("../../../../ui/__tests__/__mock__/model/tree_view");
const CLASSES = {
    filterBuilder: "dx-filterbuilder",
    groupOperation: "dx-filterbuilder-group-operation",
    treeView: "dx-treeview",
    popupWrapper: "dx-popup-wrapper",
    itemField: "dx-filterbuilder-item-field",
    itemOperation: "dx-filterbuilder-item-operation"
};
class FilterBuilderModel extends _popup.PopupModel {
    getElement() {
        const content = this.getOverlayContent();
        return null === content || void 0 === content ? void 0 : content.querySelector(`.${CLASSES.filterBuilder}`)
    }
    getGroupOperationButton() {
        const element = this.getElement();
        return element.querySelector(`.${CLASSES.groupOperation}`)
    }
    getFieldButton() {
        const element = this.getElement();
        return element.querySelector(`.${CLASSES.itemField}`)
    }
    getOperationButton() {
        const element = this.getElement();
        return element.querySelector(`.${CLASSES.itemOperation}`)
    }
    getTreeView() {
        const popups = Array.from(document.body.querySelectorAll(`.${CLASSES.popupWrapper}`));
        const treeViewPopup = popups.find((popup => popup.querySelector(`.${CLASSES.treeView}`)));
        const treeViewElement = null === treeViewPopup || void 0 === treeViewPopup ? void 0 : treeViewPopup.querySelector(`.${CLASSES.treeView}`);
        return new _tree_view.TreeViewModel(treeViewElement)
    }
}
exports.FilterBuilderModel = FilterBuilderModel;
