// devextreme/cjs/__internal/scheduler/view_model/filtration/utils/filter_by_intervals/filter_by_intervals.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/filtration/utils/filter_by_intervals/filter_by_intervals.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.filterByIntervals = void 0;
var _is_appointment_matched_intervals = require("../../../common/is_appointment_matched_intervals");
const getIntervals = (appointment, _ref) => {
    let {
        allDayIntervals: allDayIntervals,
        regularIntervals: regularIntervals,
        isDateTimeView: isDateTimeView
    } = _ref;
    if (isDateTimeView && appointment.allDay && !appointment.isAllDayPanelOccupied) {
        return regularIntervals.map((interval => ({
            min: new Date(interval.min).setUTCHours(0, 0, 0, 0),
            max: interval.max
        })))
    }
    return appointment.allDay || appointment.isAllDayPanelOccupied ? allDayIntervals : regularIntervals
};
const filterByIntervals = (appointments, options) => appointments.filter((appointment => {
    const intervals = getIntervals(appointment, options);
    const startDate = appointment.startDateUTC;
    const endDate = appointment.allDay ? appointment.endDateUTC + 1 : appointment.endDateUTC;
    return (0, _is_appointment_matched_intervals.isAppointmentMatchedIntervals)({
        startDate: startDate,
        endDate: endDate
    }, intervals)
}));
exports.filterByIntervals = filterByIntervals;
