// devextreme/cjs/__internal/scheduler/header/m_header.js
/**
 * DevExtreme (cjs/__internal/scheduler/header/m_header.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.SchedulerHeader = void 0;
require("../../../ui/drop_down_button");
var _component_registrator = _interopRequireDefault(require("../../../core/component_registrator"));
var _devices = _interopRequireDefault(require("../../../core/devices"));
var _errors = _interopRequireDefault(require("../../../core/errors"));
var _renderer = _interopRequireDefault(require("../../../core/renderer"));
var _data = require("../../../core/utils/data");
var _date = _interopRequireDefault(require("../../../core/utils/date"));
var _extend = require("../../../core/utils/extend");
var _toolbar = _interopRequireDefault(require("../../../ui/toolbar"));
var _widget = _interopRequireDefault(require("../../core/widget/widget"));
var _m_calendar = _interopRequireDefault(require("./m_calendar"));
var _m_date_navigator = require("./m_date_navigator");
var _m_utils = require("./m_utils");
var _m_view_switcher = require("./m_view_switcher");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const CLASSES = {
    component: "dx-scheduler-header",
    invisible: "dx-state-invisible"
};
const ITEM_NAMES = {
    today: "today",
    dateNavigator: "dateNavigator",
    viewSwitcher: "viewSwitcher"
};
class SchedulerHeader extends _widget.default {
    get captionText() {
        return this._getCaption().text
    }
    getIntervalOptions(date) {
        const {
            currentView: currentView,
            firstDayOfWeek: firstDayOfWeek
        } = this.option();
        const step = (0, _m_utils.getStep)(currentView.type);
        return {
            date: date,
            step: step,
            firstDayOfWeek: firstDayOfWeek,
            intervalCount: currentView.intervalCount,
            agendaDuration: currentView.agendaDuration
        }
    }
    _getDefaultOptions() {
        return (0, _extend.extend)(super._getDefaultOptions(), {
            _useShortDateFormat: !_devices.default.real().generic || _devices.default.isSimulator()
        })
    }
    _createEventMap() {
        this.eventMap = new Map([
            ["currentView", []],
            ["views", []],
            ["currentDate", [this._getCalendarOptionUpdater("value")]],
            ["min", [this._getCalendarOptionUpdater("min")]],
            ["max", [this._getCalendarOptionUpdater("max")]],
            ["tabIndex", [this.repaint.bind(this)]],
            ["focusStateEnabled", [this.repaint.bind(this)]],
            ["useDropDownViewSwitcher", [this.repaint.bind(this)]],
            ["indicatorTime", []]
        ])
    }
    _addEvent(name, event) {
        const events = this.eventMap.get(name) ?? [];
        this.eventMap.set(name, [...events, event])
    }
    _optionChanged(args) {
        const {
            name: name,
            value: value
        } = args;
        const events = this.eventMap.get(name);
        null === events || void 0 === events || events.forEach((event => {
            event(value)
        }))
    }
    onToolbarOptionChanged(fullName, value) {
        var _this$toolbar, _this$toolbar2, _this$toolbar3;
        const parts = (0, _data.getPathParts)(fullName);
        const optionName = fullName.replace(/^toolbar\./, "");
        this.option(fullName, value);
        this._toggleVisibility();
        switch (true) {
            case "toolbar" === fullName:
                this.repaint();
                break;
            case "toolbar.items" === fullName:
                null === (_this$toolbar = this.toolbar) || void 0 === _this$toolbar || _this$toolbar.option("items", value.map((item => this._parseItem(item))));
                break;
            case "items" === parts[1] && 3 === parts.length:
                null === (_this$toolbar2 = this.toolbar) || void 0 === _this$toolbar2 || _this$toolbar2.option(optionName, this._parseItem(value));
                break;
            default:
                null === (_this$toolbar3 = this.toolbar) || void 0 === _this$toolbar3 || _this$toolbar3.option(optionName, value)
        }
    }
    _init() {
        super._init();
        this._createEventMap();
        this.$element().addClass(CLASSES.component)
    }
    _render() {
        super._render();
        this._createEventMap();
        this._renderToolbar();
        this._toggleVisibility()
    }
    _renderToolbar() {
        const config = this._createToolbarConfig();
        const toolbarElement = (0, _renderer.default)("<div>");
        toolbarElement.appendTo(this.$element());
        this.toolbar = this._createComponent(toolbarElement, _toolbar.default, config)
    }
    _toggleVisibility() {
        const {
            toolbar: toolbar
        } = this.option();
        const isHeaderShown = toolbar.visible ?? toolbar.items.length;
        if (isHeaderShown) {
            this.$element().removeClass(CLASSES.invisible)
        } else {
            this.$element().addClass(CLASSES.invisible)
        }
    }
    _createToolbarConfig() {
        const {
            toolbar: toolbar
        } = this.option();
        const parsedItems = toolbar.items.map((element => this._parseItem(element)));
        return Object.assign({}, toolbar, {
            items: parsedItems
        })
    }
    _parseItem(item) {
        const itemName = "string" === typeof item ? item : item.name;
        const itemOptions = "string" === typeof item ? {} : item;
        if (itemName) {
            switch (itemName) {
                case ITEM_NAMES.today:
                    return (0, _m_date_navigator.getTodayButtonOptions)(this, itemOptions);
                case ITEM_NAMES.viewSwitcher:
                    return this.option().useDropDownViewSwitcher ? (0, _m_view_switcher.getDropDownViewSwitcher)(this, itemOptions) : (0, _m_view_switcher.getTabViewSwitcher)(this, itemOptions);
                case ITEM_NAMES.dateNavigator:
                    this._renderCalendar();
                    return (0, _m_date_navigator.getDateNavigator)(this, itemOptions);
                default:
                    _errors.default.log(`Unknown default element type: ${itemName}`)
            }
        }
        return (0, _extend.extend)(true, {}, item)
    }
    _callEvent(event, arg) {
        const events = this.eventMap.get(event);
        null === events || void 0 === events || events.forEach((EventMapHandler => EventMapHandler(arg)))
    }
    _updateCurrentView(view) {
        const {
            onCurrentViewChange: onCurrentViewChange
        } = this.option();
        onCurrentViewChange(view.name)
    }
    _updateCalendarValueAndCurrentDate(date) {
        var _this$calendar;
        this._updateCurrentDate(date);
        null === (_this$calendar = this.calendar) || void 0 === _this$calendar || _this$calendar.option("value", date)
    }
    _updateCurrentDate(date) {
        const {
            onCurrentDateChange: onCurrentDateChange
        } = this.option();
        onCurrentDateChange(date);
        this._callEvent("currentDate", date)
    }
    _renderCalendar() {
        const {
            currentDate: currentDate,
            min: min,
            max: max,
            firstDayOfWeek: firstDayOfWeek,
            focusStateEnabled: focusStateEnabled,
            tabIndex: tabIndex
        } = this.option();
        this.calendar = this._createComponent("<div>", _m_calendar.default, {
            value: currentDate,
            min: min,
            max: max,
            firstDayOfWeek: firstDayOfWeek,
            focusStateEnabled: focusStateEnabled,
            tabIndex: tabIndex,
            onValueChanged: async e => {
                var _this$calendar2;
                this._updateCurrentDate(e.value);
                await (null === (_this$calendar2 = this.calendar) || void 0 === _this$calendar2 ? void 0 : _this$calendar2.hide())
            }
        });
        this.calendar.$element().appendTo(this.$element())
    }
    _getCalendarOptionUpdater(name) {
        return value => {
            if (this.calendar) {
                this.calendar.option(name, value)
            }
        }
    }
    _getNextDate(direction, initialDate) {
        const {
            currentDate: currentDate
        } = this.option();
        const date = initialDate ?? currentDate;
        const options = this.getIntervalOptions(date);
        return (0, _m_utils.getNextIntervalDate)(options, direction)
    }
    _getDisplayedDate() {
        const {
            startViewDate: startViewDate,
            currentView: currentView
        } = this.option();
        const isMonth = "month" === currentView.type;
        return isMonth ? (0, _m_utils.nextWeek)(startViewDate) : startViewDate
    }
    _getCaptionOptions() {
        const {
            currentDate: currentDate,
            startViewDate: startViewDate
        } = this.option();
        let date = currentDate;
        if (startViewDate) {
            date = this._getDisplayedDate()
        }
        date = _date.default.trimTime(date);
        return this.getIntervalOptions(date)
    }
    _getCaption() {
        const {
            customizeDateNavigatorText: customizeDateNavigatorText
        } = this.option();
        const options = this._getCaptionOptions();
        const useShortDateFormat = this.option()._useShortDateFormat;
        return (0, _m_utils.getCaption)(options, Boolean(useShortDateFormat), customizeDateNavigatorText)
    }
    _updateDateByDirection(direction) {
        const date = this._getNextDate(direction);
        this._updateCalendarValueAndCurrentDate(date)
    }
    async _showCalendar(e) {
        var _this$calendar3;
        await (null === (_this$calendar3 = this.calendar) || void 0 === _this$calendar3 ? void 0 : _this$calendar3.show(e.element))
    }
}
exports.SchedulerHeader = SchedulerHeader;
(0, _component_registrator.default)("dxSchedulerHeader", SchedulerHeader);
