// devextreme/cjs/__internal/core/license/lcp_key_validation/lcp_key_validator.js
/**
 * DevExtreme (cjs/__internal/core/license/lcp_key_validation/lcp_key_validator.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.isProductOnlyLicense = isProductOnlyLicense;
exports.parseDevExpressProductKey = parseDevExpressProductKey;
var _const = require("../const");
var _types = require("../types");
var _const2 = require("./const");
var _license_info = require("./license_info");
var _product_info = require("./product_info");
var _utils = require("./utils");

function isProductOnlyLicense(license) {
    return "string" === typeof license && license.startsWith(_const2.LCP_SIGNATURE)
}

function productsFromString(encodedString) {
    if (!encodedString) {
        return {
            products: [],
            errorToken: _types.GENERAL_ERROR
        }
    }
    try {
        const splitInfo = encodedString.split(";");
        const productTuples = splitInfo.slice(1).filter((entry => entry.length > 0));
        const products = productTuples.map((tuple => {
            const parts = tuple.split(",");
            const version = Number.parseInt(parts[0], 10);
            const productsValue = BigInt(parts[1]);
            return (0, _product_info.createProductInfo)(version, productsValue)
        }));
        return {
            products: products
        }
    } catch (error) {
        return {
            products: [],
            errorToken: _types.DESERIALIZATION_ERROR
        }
    }
}

function parseDevExpressProductKey(productsLicenseSource) {
    if (!isProductOnlyLicense(productsLicenseSource)) {
        return _types.GENERAL_ERROR
    }
    try {
        const productsLicense = atob((0, _utils.shiftDecodeText)(productsLicenseSource.substring(_const2.LCP_SIGNATURE.length)));
        const signature = productsLicense.substring(0, _const2.SIGN_LENGTH);
        const productsPayload = productsLicense.substring(_const2.SIGN_LENGTH);
        if (!(0, _utils.verifyHash)(_const2.RSA_PUBLIC_KEY_XML, productsPayload, signature)) {
            return _types.VERIFICATION_ERROR
        }
        const {
            products: products,
            errorToken: errorToken
        } = productsFromString((0, _utils.encodeString)(productsPayload, _utils.shiftDecodeText));
        if (errorToken) {
            return errorToken
        }
        const maxVersionAllowed = (0, _license_info.findLatestDevExtremeVersion)({
            products: products
        });
        if (!maxVersionAllowed) {
            return _types.PRODUCT_KIND_ERROR
        }
        return {
            kind: _types.TokenKind.verified,
            payload: {
                customerId: "",
                maxVersionAllowed: maxVersionAllowed,
                format: _const.FORMAT
            }
        }
    } catch (error) {
        return _types.GENERAL_ERROR
    }
}
