// devextreme/cjs/__internal/core/license/lcp_key_validation/types.js
/**
 * DevExtreme (cjs/__internal/core/license/lcp_key_validation/types.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ProductKind = void 0;
var _utils = require("./utils");
const productKind = {
    Default: 0n,
    DXperienceWin: (0, _utils.bit)(0),
    XtraReports: (0, _utils.bit)(4),
    XPO: (0, _utils.bit)(15),
    DevExtremeAspNet: (0, _utils.bit)(17),
    DXperienceASP: (0, _utils.bit)(25),
    XAF: (0, _utils.bit)(28),
    Blazor: (0, _utils.bit)(31),
    DXperienceWPF: (0, _utils.bit)(38),
    DocsBasic: (0, _utils.bit)(39),
    Dashboard: (0, _utils.bit)(47),
    Snap: (0, _utils.bit)(49),
    DevExtremeHtmlJs: (0, _utils.bit)(54),
    Docs: (0, _utils.bit)(55),
    XtraReportsWpf: (0, _utils.bit)(57),
    XtraReportsWeb: (0, _utils.bit)(59),
    XtraReportsWin: (0, _utils.bit)(60),
    XtraReportsBlazor: (0, _utils.bit)(41),
    DXperienceEnt: (0, _utils.bit)(0),
    DXperienceUni: (0, _utils.bit)(0)
};
productKind.DXperienceEnt = productKind.Blazor | productKind.DXperienceWin | productKind.XtraReports | productKind.Snap | productKind.XtraReportsWin | productKind.XPO | productKind.DXperienceASP | productKind.DXperienceWPF | productKind.XtraReportsWeb | productKind.XtraReportsWpf | productKind.XtraReportsBlazor | productKind.DevExtremeAspNet | productKind.DevExtremeHtmlJs;
productKind.DXperienceUni = productKind.DXperienceEnt | productKind.XAF | productKind.DXperienceWPF | productKind.Dashboard | productKind.Docs;
const ProductKind = exports.ProductKind = Object.freeze(productKind);
