// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/helpers/utils.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/helpers/utils.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.flushAsync = exports.createDataGrid = exports.beforeTest = exports.afterTest = exports.SELECTORS = exports.GRID_CONTAINER_ID = void 0;
var _globals = require("@jest/globals");
var _fx = _interopRequireDefault(require("../../../../../../common/core/animation/fx"));
var _renderer = _interopRequireDefault(require("../../../../../../core/renderer"));
var _data_grid = _interopRequireDefault(require("../../../../../../ui/data_grid"));
var _data_grid2 = require("../../../../../grids/data_grid/__tests__/__mock__/model/data_grid");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const SELECTORS = exports.SELECTORS = {
    gridContainer: "#gridContainer"
};
const GRID_CONTAINER_ID = exports.GRID_CONTAINER_ID = "gridContainer";
const createDataGrid = async function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    return new Promise((resolve => {
        const $container = (0, _renderer.default)("<div>").attr("id", GRID_CONTAINER_ID).appendTo(document.body);
        const dataGridOptions = Object.assign({
            keyExpr: "id"
        }, options);
        const instance = new _data_grid.default($container.get(0), dataGridOptions);
        const component = new _data_grid2.DataGridModel($container.get(0));
        _globals.jest.runAllTimers();
        resolve({
            $container: $container,
            component: component,
            instance: instance
        })
    }))
};
exports.createDataGrid = createDataGrid;
const beforeTest = () => {
    _fx.default.off = true;
    _globals.jest.useFakeTimers()
};
exports.beforeTest = beforeTest;
const afterTest = () => {
    const $container = (0, _renderer.default)(SELECTORS.gridContainer);
    const dataGrid = $container.dxDataGrid("instance");
    null === dataGrid || void 0 === dataGrid || dataGrid.dispose();
    $container.remove();
    _globals.jest.clearAllMocks();
    _globals.jest.useRealTimers();
    _fx.default.off = false
};
exports.afterTest = afterTest;
const flushAsync = async () => {
    _globals.jest.runAllTimers();
    await Promise.resolve()
};
exports.flushAsync = flushAsync;
