// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/helpers/utils.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/helpers/utils.js)
 * Version: 25.2.6
 * Build date: Mon Mar 30 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    jest
} from "@jest/globals";
import fx from "../../../../../../common/core/animation/fx";
import $ from "../../../../../../core/renderer";
import DataGrid from "../../../../../../ui/data_grid";
import {
    DataGridModel
} from "../../../../../grids/data_grid/__tests__/__mock__/model/data_grid";
export const SELECTORS = {
    gridContainer: "#gridContainer"
};
export const GRID_CONTAINER_ID = "gridContainer";
export const createDataGrid = async function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    return new Promise((resolve => {
        const $container = $("<div>").attr("id", "gridContainer").appendTo(document.body);
        const dataGridOptions = Object.assign({
            keyExpr: "id"
        }, options);
        const instance = new DataGrid($container.get(0), dataGridOptions);
        const component = new DataGridModel($container.get(0));
        jest.runAllTimers();
        resolve({
            $container: $container,
            component: component,
            instance: instance
        })
    }))
};
export const beforeTest = () => {
    fx.off = true;
    jest.useFakeTimers()
};
export const afterTest = () => {
    const $container = $(SELECTORS.gridContainer);
    const dataGrid = $container.dxDataGrid("instance");
    null === dataGrid || void 0 === dataGrid || dataGrid.dispose();
    $container.remove();
    jest.clearAllMocks();
    jest.useRealTimers();
    fx.off = false
};
export const flushAsync = async () => {
    jest.runAllTimers();
    await Promise.resolve()
};
